### PR TITLE
Port Instructions to meta ; part 3, Instruction/InstructionGroup + definitions

### PR DESCRIPTION
**Pull request by bnjbvr**
Based off https://github.com/CraneStation/cranelift/pull/706

### CODE REVIEW

File: cranelift-codegen/meta/src/cdsl/inst.rs [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+impl Instruction {
+    pub fn snake_name(&self) -> &'static str {
+        if self.name == "return" {
+            "return_"
+        } else {
+            self.name
+        }
+    }
+
+    // XXX was named blurb
```
Comment (jgouly) [inline cranelift-codegen/meta/src/cdsl/inst.rs]: I guess this can be removed?

File: cranelift-codegen/meta/src/cdsl/inst.rs [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+    pub fn doc_comment_first_line(&self) -> &'static str {
+        for line in self.doc.split("\n") {
+            let stripped = line.trim();
+            if stripped.len() > 0 {
+                return stripped;
+            }
+        }
+        return "";
```
Comment (jgouly) [inline cranelift-codegen/meta/src/cdsl/inst.rs]: Just replace with ""

File: cranelift-codegen/meta/src/cdsl/inst.rs [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+            can_load: false,
+            can_store: false,
+            can_trap: false,
+            other_side_effects: false,
+        }
+    }
+
+    pub fn operands_in(mut self, operands: Vec<&Operand>) -> Self {
+        assert!(self.operands_in.is_none());
+        self.operands_in = Some(operands.iter().map(|x| (*x).clone()).collect());
```
Comment (jgouly) [inline cranelift-codegen/meta/src/cdsl/inst.rs]: Can you use .cloned() here? (and below)

**Comment by jgouly** [comment]
LGTM too, just a few very minor comments. I also didn't do a deep review of `cranelift-codegen/meta/src/shared/instructions.rs`, just randomly looked at some of the instructions. Hopefully tests will catch issues if there are any!

File: cranelift-codegen/meta/src/cdsl/inst.rs (hunk 3 above)
Comment (bnjbvr) [reply cranelift-codegen/meta/src/cdsl/inst.rs]: No, because the parameter `operands` is a `Vec<&Operand>` while the `operands_in` is a `Vec<Operand>`.

**Comment by bnjbvr** [comment]
Thanks for the reviews!

**Pull request merged**